web3/middleware/cache.py

```diff
@@ -0,0 +1,359 @@
+    'eth_getLogs',
+    # 'eth_getWork',
+    # 'eth_submitWork',
+    # 'eth_submitHashrate',
+}
+
+
+AVG_BLOCK_TIME_KEY = 'avg_block_time'
+AVG_BLOCK_SAMPLE_SIZE_KEY = 'avg_block_sample_size'
+AVG_BLOCK_TIME_UPDATED_AT_KEY = 'avg_block_time_updated_at'
+
+
+def should_cache_by_block_number(method, params, response):
+    if method == 'eth_getBlockByNumber':
+        if params == ['latest'] or params == ['pending']:
+            return False
+
+    if 'error' in response:
+        return False
+    elif 'result' not in response:
+        return False
+
+    if response['result'] is None:
+        return False
+    return True
+
+
+def construct_block_number_based_cache_middleware(cache,
+                                                  rpc_whitelist=BLOCK_NUMBER_RPC_WHITELIST,
+                                                  average_block_time_sample_size=240,
+                                                  default_average_block_time=15,
+                                                  should_cache_fn=should_cache_by_block_number):
+    """
+    Constructs a middleware which caches responses based on the request
+    ``method``, ``params``, and the current latest block hash.
+
+    :param cache: Any dictionary-like object
+    :param cache_expire_seconds: The number of seconds an item may be cached before it should expire.
+    :param rpc_whitelist: A set of RPC methods which may have their responses cached.
+    :param should_cache_fn: A callable which accepts ``method`` ``params`` and ``response`` and returns a boolean as to whether the response should be cached.
+
+    .. note::
+        This middleware avoids re-fetching the current latest block for each
+        request by tracking the current average block time and only requesting
+        a new block when the last seen latest block is older than the average
+        block time.
+    """
+    def block_number_based_cache_middleware(make_request, web3):
+        block_info = {}
+
+        def _update_block_info_cache(web3):
+            avg_block_time = block_info.get(AVG_BLOCK_TIME_KEY, default_average_block_time)
+            avg_block_sample_size = block_info.get(AVG_BLOCK_SAMPLE_SIZE_KEY, 0)
+            avg_block_time_updated_at = block_info.get(AVG_BLOCK_TIME_UPDATED_AT_KEY, 0)
+
+            # compute age as counted by number of blocks since the avg_block_time
+            if avg_block_time == 0:
+                avg_block_time_age_in_blocks = avg_block_sample_size
+            else:
+                avg_block_time_age_in_blocks = (
+                    (time.time() - avg_block_time_updated_at) / avg_block_time
+                )
+
+            if avg_block_time_age_in_blocks >= avg_block_sample_size:
+                # If the length of time since the average block time as
+                # measured by blocks is greater than or equal to the number of
+                # blocks sampled then we need to recompute the average block
+                # time.
```

Comment:
To me, this was harder to read to have everything in blocks instead of time.

For example, `cache_expiration_seconds` seems more natural than `avg_block_sample_size`, and leads to a test that probably doesn't require explanation, like:

```py
if time.time() - avg_block_time_updated_at >= cache_expiration_seconds
```

It seems worth the cost of having to do a bit more calculation to get the ancestor block number for the next sample.

Reply:
I wrote it this way to cover the case where there is not enough chain history to provide an accurate average block time.

With your formula, the avg time would get set to `block_1_timestamp - genesis_timestamp` if queried on the 1st block on the chain.  Then it wouldn't get refreshed until `cache_expiration_seconds` later.

The current code would result in it getting updated at approximately blocks 1, 2, 4, 8, 16, ... until the chain history was at least `cache_expiration_seconds` long after which if would update at regular intervals.

Thoughts on whether the complexity is worth covering this admittedly minority use case?

Reply:
Doesn't seem worth it to me. I'd rather see a hard-coded corner case that is easy to read in code (and easy to explain in documentation), than to complicate the main body.

**But** with the included commentary, and the fact that it's a non-default feature, I'm happy to merge now and come back to it later.